### PR TITLE
Generalize the Consensus Infrastructure

### PR DESCRIPTION
**Pull request by gnunicorn**
In our efforts to allow pluggable consensus engines in substrate, this PR attempts to generalize consensus messages over the network, blocks and the client.

 - [x] Generalize `NetworkMessage`s for Consensus as an opaque `Vec<u8>`
 - [x] Generalize `Justification` as a Consensus-specific opaque `Vec<u8>`
 - [x] Generalize `ConsensusGossip` over any consensus internal information
 - [x] Remove all rhododendron-dependencies (and everything `bft:*`) from `core`, `sr-*` and the `srml`
 - [x] Restructure Import-Procedure as described in #784
    - [x] Decouple `import_block` from `client`
    - [x] generalize client over `cli` to allow substrate to decide on consensus engine
    - [x] let [allow cli to block-import again](https://github.com/paritytech/substrate/issues/784#issuecomment-427453215)
 - [x] disable current rhd-code-base; make it all compile

### CODE REVIEW

**Comment by gavofyork** [comment]
> because GRANDPA and AURA have a higher priority at the moment, we might post-pone fixing rhd for later and merge this after the restructure itself finished, breaking consensus for node on master...

tht's fine - bbq birch will be decommissioned as soon as shaft(grandpa)/aura are operational.

**Comment by gnunicorn** [comment]
Updated the comment to reflect the latest design and process decisions, specifically: 
 - fix block-import for cli by providing allow substrate/node to provide a specialization to decide on consensus engines
 - drop attempt to fix RHD; instead disable and focus on getting CI pass without it for now 


---

Previously, this also attempted to, but won't be part of this 
> - [ ] Re-activate rhododendron (`substrate-consensus-rhd`)\*
>   - [ ] simplify message handling [by directly supporting for parity-codec](https://github.com/paritytech/rhododendron/pull/17)
>   - [x] move `MisbehaviourCheck` into `rhd`-crate (as it is `rhd` specific)
>   - [ ] refactor rhodendron to use new `ImportQueue`-System
>   - [ ] re-implement rhd in `node`
>   - [ ] ensure `rhd`-messages are backwards-compatible

File: core/network/src/import_queue.rs [hunk 1]
```diff
@@ -499,33 +511,68 @@ unsafe impl<B: BlockT> Send for ImportCB<B> {}
...
+			finalized: self.0,
+			external_justification: justification,
+			internal_justification: vec![],
+			auxiliary: Vec::new(),
+		}, None))
+	}
+}
+
+#[cfg(any(test, feature = "test-helpers"))]
+unsafe impl Send for PassThroughVerifier {}
```
Comment (rphmeier) [inline core/network/src/import_queue.rs]: `PassThroughVerifier` stores only a bool -- are these necessary?

File: core/consensus/rhd/src/lib.rs [hunk 2]
```diff
@@ -616,110 +632,118 @@ pub fn bft_threshold(n: usize) -> usize {
...
-	if ed25519::verify_strong(&sig.signature, &message, &sig.signer) {
-		Ok(())
-	} else {
-		Err(ErrorKind::InvalidSignature(sig.signature.into(), sig.signer.clone().into()).into())
-	}
-}
-
-/// Sign a BFT message with the given key.
-pub fn sign_message<B: Block + Clone>(message: Message<B>, key: &ed25519::Pair, parent_hash: B::Hash) -> LocalizedMessage<B> {
+// fn check_justification_signed_message<H>(
```
Comment (bkchr) [inline core/consensus/rhd/src/lib.rs]: Why did you need this code entirely? If you need this code anytime in the future, you can just use git to get it back^^

File: core/client/src/client.rs [hunk 3]
```diff
@@ -486,37 +518,24 @@ impl<B, E, Block> Client<B, E, Block> where
...
-	/// Queue a block for import.
+	/// Import a checked and validated block
 	pub fn import_block(
 		&self,
-		origin: BlockOrigin,
-		header: JustifiedHeader<Block>,
-		body: Option<Vec<<Block as BlockT>::Extrinsic>>,
-		finalized: bool,
+		import_block: ImportBlock<Block>,
+	    new_authorities: Option<Vec<AuthorityId>>,
```
[the remark this answers is not in the document]
Comment (gnunicorn) [reply core/client/src/client.rs]: @svyatonik the meaning has changed. While you described how it was previously, all actual checks are done within the `Verifier` now and this only communicates the changes to import: the Block itself and optionally a new Set of Authorities - as changes of authority set will also be consensus engine driven. This is a temporary hack until we have the signaling in place, in which case this wouldn't be passed here anymore but instead a `NewAuthorities`-Signal will be triggered at import time instead.

[the remark this answers is not in the document]
Comment (svyatonik) [reply core/client/src/client.rs]: OK, thanks. Then better to wait for actual `Verifier` implementation && see if some changes are required here (and in the authorities cache).

File: node/service/src/lib.rs [hunk 4]
```diff
@@ -71,10 +60,33 @@ pub type ChainSpec = service::ChainSpec<GenesisConfig>;
 pub type ComponentClient<C> = Client<<C as Components>::Backend, <C as Components>::Executor, Block>;
 pub type NetworkService = network::Service<Block, <Factory as service::ServiceFactory>::NetworkProtocol, Hash>;
 
+/// A verifier that doesn't actually do any checks
+pub struct NoneVerifier;
```
Comment (gavofyork) [inline node/service/src/lib.rs]: Using `()` for "null" implementations of traits by convention.

Comment (gnunicorn) [reply node/service/src/lib.rs]: Can't be done here: `impl<B: BlockT> Verifier<B> for ()` can only be done within the crate that defines `Verifier`. However we don't want to provide such a bad implementation in `substrate-network`.

Leaving it as is for now, as this is only temporarily to make the compiler happy and will be replaced in the next chunk of work with an actual `Verifier` implementation.

**Pull request merged**